File: apps/web/lib/books.ts
export type Book = {
  id: string;
  title: string;
  author: string;
  years: string;
  era: string;
  theme: string;
  description: string;
  coverTone: string;
  coverInk: string;
  coverMark: string;
  cover?: { imageUrl: string; sourceUrl: string; label: string; credit: string; license: string; licenseUrl: string; checkedAt: string };
  startYear: number | null;
  endYear: number | null;
  approximate: boolean;
  creators: { id: string; name: string; description: string; birth: string | null; death: string | null; sourceUrl: string; kind?: string; credit?: string }[];
  sourceUrl: string;
  dateBasis: string;
  selectionBasis: string;
  status: string;
};

export type BookRange = { start: number; end: number };
export type BookFilterOption = { slug: string; name: string };
export type BooksFacets = { languages: BookFilterOption[]; countries: BookFilterOption[]; regions: BookFilterOption[] };
export type BookSuggestion = { key: "language" | "country" | "region" | "author"; value: string; name: string; count: number };
export type TimelineAuthor = Book["creators"][number] & {
  startYear: number | null; endYear: number | null; lifespan: string; approximate: boolean; bookCount: number; credits: string[];
};
export type BooksResponse = {
  items: Book[];
  total: number;
  selectionTotal: number;
  hasMore: boolean;
  nextCursor: string;
  range: BookRange;
  bounds: BookRange;
  undatedTotal: number;
  mode: "individual" | "density";
  density: { start_year: number; end_year: number; count: number }[];
  suggested_filters?: BookSuggestion[];
  authors?: TimelineAuthor[];
  view?: "books" | "authors";
  periods: (BookRange & { label: string })[];
  ticks: { year: number; label: string }[];
};

export function bookYearLabel(year: number): string {
  return year < 0 ? `${Math.abs(year)} BCE` : `${year}`;
}

// Format the server's lifespan envelope compactly on the chart. The drawer
// retains the original birth/death labels, including alternatives and ranges.
export function authorLifespanLabel(author: TimelineAuthor): string {
  if (author.birth && author.death && author.startYear !== null && author.endYear !== null) {
    return `${author.approximate ? "c. " : ""}${bookYearLabel(author.startYear)}–${bookYearLabel(author.endYear)}`;
  }
  return author.lifespan;
}

export function compressedBCE(range: BookRange): boolean { return range.start < 1 && range.end > 1; }

const calendar = (year: number) => year < 0 ? year + 1 : year;
const historicalYear = (value: number) => value <= 0 ? value - 1 : value;
export function compressedBefore1700(range: BookRange): boolean { return range.start < 1700 && range.end > 1700; }

// Fixed historical axis: 5000 BCE / 1 / linearFrom / 2000 anchor 0 / 5 / 25 / 100.
// Books uses 1700; All uses 1400. Selecting years never rescales these anchors.
function bookCoordinate(year: number, linearFrom: number): number {
 const y=calendar(year);
 return y<=1 ? (y+4999)/5000*5 : y<=linearFrom ? 5+(y-1)/(linearFrom-1)*20 : 25+(y-linearFrom)/(2000-linearFrom)*75;
}
function bookCoordinateYear(value: number, linearFrom: number): number {
 const y=value<=5 ? value/5*5000-4999 : value<=25 ? 1+(value-5)/20*(linearFrom-1) : linearFrom+(value-25)/75*(2000-linearFrom);
 return historicalYear(Math.round(y));
}
export function bookTickPosition(year: number, range: BookRange, linearFrom = 1700): number {
 return (bookCoordinate(year,linearFrom)-bookCoordinate(range.start,linearFrom))/(bookCoordinate(range.end,linearFrom)-bookCoordinate(range.start,linearFrom))*100;
}
export function bookYearAtPosition(position: number, range: BookRange, linearFrom = 1700): number {
 const fraction=Math.max(0,Math.min(100,position))/100;
 return bookCoordinateYear(bookCoordinate(range.start,linearFrom)+fraction*(bookCoordinate(range.end,linearFrom)-bookCoordinate(range.start,linearFrom)),linearFrom);
}

// Tick visibility follows pixel spacing, since equal calendar steps are no
// longer equally spaced. Keep the change of scale and both ends legible.
export function visibleBookTicks(ticks: BooksResponse["ticks"], range: BookRange, width: number, linearFrom = 1700) {
  if (!ticks.length) return [];
  const compressed = range.start < linearFrom && range.end > linearFrom;
  const candidates = [...ticks];
  if (compressed && !candidates.some(tick => tick.year === linearFrom)) candidates.push({ year: linearFrom, label: String(linearFrom) });
  candidates.sort((a, b) => a.year - b.year);
  const chosen = [candidates[0], candidates[candidates.length - 1]];
  const boundary = candidates.find(tick => tick.year === linearFrom);
  if (compressed && boundary) chosen.push(boundary);
  for (const tick of candidates) {
    if (chosen.every(other => Math.abs(bookTickPosition(tick.year, range, linearFrom) - bookTickPosition(other.year, range, linearFrom)) * width / 100 >= 70)) chosen.push(tick);
  }
  return [...new Map(chosen.map(tick => [tick.year, tick])).values()].sort((a, b) => a.year - b.year);
}

// Axis labels are presentation only. Date selection still belongs to the API,
// while the complete historical axis stays fixed across range changes.
export function bookAxisTicks(bounds: BookRange, width: number, linearFrom = 1700) {
  const ticks = [{ year: Math.round(bounds.start / 2), label: "BCE" }];
  for (let year = 100; year <= bounds.end; year += 100) ticks.push({ year, label: String(year) });
  return visibleBookTicks(ticks, bounds, width, linearFrom);
}

// Visual layout only; the API owns chronology, filtering and page order.
export function positionBooks<T extends { startYear: number | null; endYear: number | null }>(books: T[], range: BookRange, width: number, linearFrom = 1700) {
  const pixels = Math.max(1, width);
  const lanes: [number, number][][] = [];
  return books.filter((book): book is T & { startYear: number; endYear: number } => book.startYear !== null && book.endYear !== null).map(book => {
    const x = bookTickPosition(Math.max(range.start, book.startYear), range, linearFrom) / 100 * pixels;
    const end = bookTickPosition(Math.min(range.end, book.endYear), range, linearFrom) / 100 * pixels;
    const markWidth = Math.min(pixels, Math.max(8, end - x));
    const left = Math.min(x, pixels - markWidth);
    const labelWidth = Math.min(pixels, 246);
    const labelLeft = Math.max(0, Math.min(left, pixels - labelWidth));
    const occupied: [number, number] = [Math.min(left, labelLeft), Math.max(left + markWidth, labelLeft + labelWidth)];
    let lane = lanes.findIndex(intervals => intervals.every(([a, b]) => occupied[0] >= b + 18 || occupied[1] + 18 <= a));
    if (lane < 0) { lane = lanes.length; lanes.push([]); }
    lanes[lane].push(occupied);
    return { ...book, lane, left: left / pixels * 100, width: markWidth / pixels * 100, labelOffset: labelLeft - left, labelWidth };
  });
}

File: apps/server/internal/books/books.go
// Package books owns catalogue visibility, chronology and bounded reads.
package books

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

type Creator struct {
	ID             string  `json:"id"`
	Name           string  `json:"name"`
	Description    string  `json:"description"`
	Birth          *string `json:"birth"`
	Death          *string `json:"death"`
	SourceURL      string  `json:"sourceUrl"`
	SourceRevision *int64  `json:"sourceRevision"`
	Kind           string  `json:"kind"`
	Credit         string  `json:"credit"`
}
type Book struct {
	ID             string    `json:"id"`
	SourceID       string    `json:"sourceId"`
	Title          string    `json:"title"`
	Author         string    `json:"author"`
	Years          string    `json:"years"`
	Era            string    `json:"era"`
	Theme          string    `json:"theme"`
	Description    string    `json:"description"`
	CoverTone      string    `json:"coverTone"`
	CoverInk       string    `json:"coverInk"`
	CoverMark      string    `json:"coverMark"`
	Cover          *Cover    `json:"cover,omitempty"`
	StartYear      *int      `json:"startYear"`
	EndYear        *int      `json:"endYear"`
	Approximate    bool      `json:"approximate"`
	Creators       []Creator `json:"creators"`
	SourceURL      string    `json:"sourceUrl"`
	SourceRevision *int64    `json:"sourceRevision"`
	SelectionBasis string    `json:"selectionBasis"`
	DateBasis      string    `json:"dateBasis"`
	Status         string    `json:"status"`
}
type Range struct {
	Start int `json:"start"`
	End   int `json:"end"`
}
type Period struct {
	Range
	Label string `json:"label"`
}
type DensityPeriod struct {
	Start int `json:"start_year"`
	End   int `json:"end_year"`
	Count int `json:"count"`
}
type Tick struct {
	Year  int    `json:"year"`
	Label string `json:"label"`
}
type Filter struct {
	Range
	Query, After                  string
	Authors                       []string
	Languages, Countries, Regions []string
	Women, Top100                 bool
	Limit                         int
	Preview                       bool
	View                          string
}
type FilterOption struct {
	Slug string `json:"slug"`
	Name string `json:"name"`
}
type SuggestedFilter struct {
	Key   string `json:"key"`
	Value string `json:"value"`
	Name  string `json:"name"`
	Count int    `json:"count"`
}
type Facets struct {
	Languages []FilterOption `json:"languages"`
	Countries []FilterOption `json:"countries"`
	Regions   []FilterOption `json:"regions"`
}
type Response struct {
	Items            []Book            `json:"items"`
	Total            int               `json:"total"`
	SelectionTotal   int               `json:"selectionTotal"`
	UndatedTotal     int               `json:"undatedTotal"`
	HasMore          bool              `json:"hasMore"`
	NextCursor       string            `json:"nextCursor"`
	Range            Range             `json:"range"`
	Bounds           Range             `json:"bounds"`
	Periods          []Period          `json:"periods"`
	Ticks            []Tick            `json:"ticks"`
	Mode             string            `json:"mode"`
	Density          []DensityPeriod   `json:"density"`
	SuggestedFilters []SuggestedFilter `json:"suggested_filters"`
	Authors          []TimelineAuthor  `json:"authors,omitempty"`
	View             string            `json:"view,omitempty"`
}
type AuthorOptions struct {
	Items   []string `json:"items"`
	HasMore bool     `json:"hasMore"`
}

var Bounds = Range{-5000, 2000}
var ErrNotFound = errors.New("book not found")
var ErrUnavailable = errors.New("book catalogue unavailable")

func (f Filter) Validate() error {
	if f.View != "" && f.View != "books" && f.View != "authors" {
		return fmt.Errorf("choose view=books or view=authors")
	}
	if f.Start < Bounds.Start || f.End > Bounds.End || f.Start >= f.End || f.Start == 0 || f.End == 0 {
		return fmt.Errorf("choose an ordered range from 5000 BCE to 2000; there is no year zero")
	}
	if len(f.Query) > 200 || len(f.After) > 512 {
		return fmt.Errorf("book search or cursor is too long")
	}
	if f.Limit < 1 || f.Limit > 100 {
		return fmt.Errorf("limit must be from 1 to 100")
	}
	for _, choices := range [][]string{f.Authors, f.Languages, f.Countries, f.Regions} {
		if len(choices) > 32 {
			return fmt.Errorf("choose up to 32 values per filter")
		}
		for _, choice := range choices {
			if len(choice) > 250 || strings.TrimSpace(choice) == "" {
				return fmt.Errorf("invalid book filter value")
			}
		}
	}
	_, err := decodeCursor(f.After)
	return err
}

type cursor struct {
	Year int    `json:"year"`
	ID   string `json:"id"`
}

func decodeCursor(raw string) (cursor, error) {
	c := cursor{Year: -5001}
	if raw == "" {
		return c, nil
	}
	b, err := base64.RawURLEncoding.DecodeString(raw)
	if err != nil || json.Unmarshal(b, &c) != nil || c.ID == "" || len(c.ID) > 100 || c.Year == 0 || c.Year < Bounds.Start || (c.Year > Bounds.End && c.Year != 2147483647) {
		return c, fmt.Errorf("invalid book cursor")
	}
	return c, nil
}
func encodeCursor(book Book) string {
	year := 2147483647
	if book.StartYear != nil {
		year = *book.StartYear
	}
	b, _ := json.Marshal(cursor{year, book.ID})
	return base64.RawURLEncoding.EncodeToString(b)
}
func YearLabel(year int) string {
	if year < 0 {
		return fmt.Sprintf("%d BCE", -year)
	}
	return fmt.Sprint(year)
}

// Same density intervals as ArtWorks: 50, 25, then 10 years. In a mixed
// overview all BCE years occupy one explicitly labelled opening interval.
func densityPeriods(r Range) []DensityPeriod {
	width := 10
	if r.End-r.Start > 200 {
		width = 25
	}
	if r.End-r.Start > 500 {
		width = 50
	}
	periods := []DensityPeriod{}
	start := r.Start
	if start < 0 && r.End > 0 {
		periods = append(periods, DensityPeriod{start, -1, 0})
		start = 1
	}
	for start <= r.End {
		end := start + width - 1
		if start == 1 {
			end = width - 1
		}
		if end >= r.End-1 {
			end = r.End
		}
		if end == 0 {
			end = -1
		}
		periods = append(periods, DensityPeriod{start, end, 0})
		start = end + 1
		if start == 0 {
			start = 1
		}
	}
	return periods
}
func metadata(r Range) Response {
	presets := []Period{{Bounds, "Full range"}, {Range{Bounds.Start, -1}, "BCE"}, {Range{1, Bounds.End}, "1–2000"}}
	for _, span := range []int{100, 50, 25} {
		start := max(Bounds.Start, min(Bounds.End-span, (r.Start+r.End-span)/2))
		end := start + span
		if start == 0 {
			start = 1
		}
		if end == 0 {
			end = 1
		}
		label := fmt.Sprintf("%d years", span)
		if span == 100 {
			label = "Century"
		}
		presets = append(presets, Period{Range{start, end}, label})
	}
	result := Response{Items: []Book{}, Range: r, Bounds: Bounds, Periods: presets, Density: []DensityPeriod{}, Ticks: []Tick{}, Mode: "individual"}
	span := r.End - r.Start
	tickStart := r.Start
	if r.Start < 0 && r.End > 500 {
		span = r.End
		tickStart = 1
		result.Ticks = append(result.Ticks, Tick{r.Start / 2, "BCE"})
	}
	step := 1
	switch {
	case span > 500:
		step = 100
	case span > 200:
		step = 50
	case span > 80:
		step = 25
	case span > 35:
		step = 10
	case span > 10:
		step = 5
	}
	for year := tickStart; year <= r.End; year++ {
		if year != 0 && year%step == 0 {
			result.Ticks = append(result.Ticks, Tick{year, YearLabel(year)})
		}
	}
	return result
}
